cdap-data-fabric/src/main/java/co/cask/cdap/store/OwnerStore.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.store;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.TxRunnable;
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.dataset.DatasetManagementException;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.table.ConflictDetection;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.common.AlreadyExistsException;
+import co.cask.cdap.common.kerberos.SecurityUtil;
+import co.cask.cdap.data.dataset.SystemDatasetInstantiator;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.MultiThreadDatasetCache;
+import co.cask.cdap.data2.transaction.Transactions;
+import co.cask.cdap.data2.transaction.TxCallable;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.NamespaceId;
+import com.google.inject.Inject;
+import org.apache.tephra.RetryStrategies;
+import org.apache.tephra.TransactionFailureException;
+import org.apache.tephra.TransactionSystemClient;
+
+import java.io.IOException;
+import java.util.Collections;
+
+/**
+ * This class manages owner's principal information of CDAP entities.
+ * <p>
+ * Currently: Owner information is stored for the following entities:
+ * <ul>
+ * <li>{@link co.cask.cdap.api.data.stream.Stream}</li>
+ * <li>{@link co.cask.cdap.api.dataset.Dataset}</li>
+ * <li>{@link co.cask.cdap.api.app.Application}</li>
+ * </ul>
+ * </p>
+ * <p>
+ * It is the responsibility of the creator of the supported entities to add an entry in this store to store the
+ * associated owner's principal. Note: An absence of an entry in this table for an {@link EntityId} does not
+ * signifies that the entity does not exists. The owner information is only stored if an owner was provided during
```

Comment:
I agree with keeping the provided owner but I also don't want the code to get the name if one was not given spread throughout the codebase. What if that changes and say we want to assign cdap the owner to all the entities that don't have an owner. We will have to find all the places it is used and change. Also, I don't think the second point of confusion is right. With impersonation turned on every entity is going to be impersonated and every entity will have an owner, so I think seeing who the owner is, whether set explicitly or implicitly, would be more important. Let us discuss this tomorrow and see if we can come up with an approach that satisfies both requirements.